tstest/integration/nat: TestEasyEasy flake

https://github.com/tailscale/tailscale/actions/runs/22749035043/job/65979408358?pr=18905
```
exec /home/runner/.cache/tsgo/5b5cb0db47535a0a8d2f450cb1bf83af8e70f164/bin/go test -v '-run=^TestEasyEasy$' -timeout=3m -count=1 ./tstest/integration/nat --run-vm-tests
go: downloading golang.org/x/mod v0.31.0
go: downloading golang.org/x/sync v0.19.0
go: downloading github.com/go-json-experiment/json v0.0.0-20250813024750-ebf49471dced
go: downloading golang.org/x/sys v0.40.0
go: downloading github.com/gaissmai/bart v0.26.1
go: downloading github.com/google/gopacket v1.1.19
go: downloading go4.org/mem v0.0.0-20240501181205-ae6ca9944745
go: downloading gvisor.dev/gvisor v0.0.0-20260224225140-573d5e7127a8
go: downloading github.com/fxamacker/cbor/v2 v2.9.0
go: downloading github.com/hdevalence/ed25519consensus v0.2.0
go: downloading golang.org/x/crypto v0.46.0
go: downloading golang.org/x/net v0.48.0
go: downloading github.com/google/btree v1.1.3
go: downloading golang.org/x/time v0.12.0
go: downloading github.com/axiomhq/hyperloglog v0.0.0-20240319100328-84253e514e02
go: downloading github.com/klauspost/compress v1.18.2
go: downloading golang.org/x/exp v0.0.0-20250620022241-b7579e27df2b
go: downloading go4.org/netipx v0.0.0-20231129151722-fdeea329fbba
go: downloading github.com/mitchellh/go-ps v1.0.0
go: downloading github.com/coder/websocket v1.8.12
go: downloading filippo.io/edwards25519 v1.2.0
go: downloading github.com/x448/float16 v0.8.4
go: downloading github.com/dgryski/go-metro v0.0.0-20180109044635-280f6062b5bc
go: downloading github.com/jsimonetti/rtnetlink v1.4.0
go: downloading github.com/mdlayher/netlink v1.7.3-0.20250113171957-fbb4dce95f42
go: downloading golang.org/x/text v0.32.0
go: downloading github.com/mdlayher/socket v0.5.0
=== RUN   TestEasyEasy
    nat_test.go:72: building VM image...
go run build.go --build --app=natlabapp
go: downloading github.com/bradfitz/monogok v0.0.0-20260208031948-2219c393d032
go: downloading github.com/spf13/cobra v1.10.2
go: downloading golang.org/x/crypto/x509roots/fallback v0.0.0-20260113154411-7d0074ccc6f1
go: downloading github.com/mattn/go-isatty v0.0.20
go: downloading github.com/spf13/pflag v1.0.9
2026/03/06 04:27:15 monogok g2219c3 on GOARCH=amd64 GOOS=linux
2026/03/06 04:27:15 
2026/03/06 04:27:15 Build target: GOOS=linux GOARCH=amd64 GOARCH=amd64 GOOS=linux
2026/03/06 04:27:15 Build timestamp: 2026-03-06T04:27:15Z
2026/03/06 04:27:15 Loading system CA certificates from /etc/ssl/certs/ca-certificates.crt
2026/03/06 04:27:15 Building 4 Go packages:
2026/03/06 04:27:15 
2026/03/06 04:27:15   github.com/gokrazy/serial-busybox
2026/03/06 04:27:15 
2026/03/06 04:27:15   tailscale.com/cmd/tailscale
2026/03/06 04:27:15 
2026/03/06 04:27:15   tailscale.com/cmd/tailscaled
2026/03/06 04:27:15 
2026/03/06 04:27:15   tailscale.com/cmd/tta
2026/03/06 04:27:15 
go: downloading github.com/gokrazy/gokrazy v0.0.0-20260123094004-294c93fa173c
go: downloading github.com/google/renameio/v2 v2.0.0
go: downloading github.com/mdlayher/packet v1.1.2
go: downloading github.com/rtr7/dhcp4 v0.0.0-20220302171438-18c84d089b46
go: downloading github.com/vishvananda/netlink v1.3.1-0.20240922070040-084abd93d350
go: downloading github.com/josharian/native v1.1.0
go: downloading github.com/vishvananda/netns v0.0.5
go: downloading github.com/gokrazy/serial-busybox v0.0.0-20250119153030-ac58ba7574e7
go: downloading github.com/gokrazy/internal v0.0.0-20251208203110-3c1aa9087c82
go: downloading github.com/kenshaw/evdev v0.1.0
go: downloading github.com/mdlayher/watchdog v0.0.0-20221003142519-49be0df7b3b5
go: downloading github.com/kballard/go-shellquote v0.0.0-20180428030007-95032a82bc51
go: downloading github.com/mattn/go-colorable v0.1.13
go: downloading github.com/peterbourgon/ff/v3 v3.4.0
go: downloading github.com/toqueteos/webbrowser v1.2.0
go: downloading k8s.io/client-go v0.34.0
go: downloading sigs.k8s.io/yaml v1.6.0
go: downloading software.sslmate.com/src/go-pkcs12 v0.4.0
go: downloading fyne.io/systray v1.11.1-0.20250812065214-4856ac3adc3c
go: downloading github.com/Kodeworks/golang-image-ico v0.0.0-20141118225523-73f0f4cfade9
go: downloading github.com/atotto/clipboard v0.1.4
go: downloading github.com/fogleman/gg v1.3.0
go: downloading github.com/godbus/dbus/v5 v5.1.1-0.20230522191255-76236955d466
go: downloading github.com/tailscale/web-client-prebuilt v0.0.0-20250124233751-d4cd19a26976
go: downloading go.yaml.in/yaml/v2 v2.4.2
go: downloading github.com/tailscale/hujson v0.0.0-20221223112325-20486734a56a
go: downloading github.com/golang/groupcache v0.0.0-20241129210726-2c02b8208cf8
go: downloading github.com/skip2/go-qrcode v0.0.0-20200617195104-da1b6568686e
go: downloading github.com/golang/freetype v0.0.0-20170609003504-e2365dfdc4a0
go: downloading golang.org/x/image v0.27.0
go: downloading golang.org/x/oauth2 v0.33.0
go: downloading github.com/aws/aws-sdk-go-v2 v1.41.0
go: downloading github.com/aws/aws-sdk-go-v2/config v1.29.5
go: downloading github.com/aws/aws-sdk-go-v2/service/ssm v1.44.7
go: downloading github.com/aws/aws-sdk-go-v2/feature/ec2/imds v1.16.27
go: downloading github.com/aws/aws-sdk-go-v2/service/sts v1.41.5
go: downloading github.com/aws/smithy-go v1.24.0
go: downloading github.com/huin/goupnp v1.3.0
go: downloading github.com/aws/aws-sdk-go-v2/credentials v1.17.58
go: downloading github.com/aws/aws-sdk-go-v2/internal/ini v1.8.2
go: downloading github.com/aws/aws-sdk-go-v2/service/sso v1.24.14
go: downloading github.com/aws/aws-sdk-go-v2/service/ssooidc v1.28.13
go: downloading github.com/aws/aws-sdk-go-v2/internal/configsources v1.4.16
go: downloading github.com/aws/aws-sdk-go-v2/service/internal/accept-encoding v1.13.4
go: downloading github.com/aws/aws-sdk-go-v2/service/internal/presigned-url v1.13.16
go: downloading github.com/jmespath/go-jmespath v0.4.0
go: downloading github.com/aws/aws-sdk-go-v2/internal/endpoints/v2 v2.7.16
go: downloading github.com/djherbis/times v1.6.0
go: downloading github.com/tailscale/xnet v0.0.0-20240729143630-8497ac4dab2e
go: downloading golang.org/x/term v0.38.0
go: downloading github.com/creachadair/msync v0.7.1
go: downloading github.com/creack/pty v1.1.24
go: downloading github.com/pkg/sftp v1.13.6
go: downloading github.com/u-root/u-root v0.14.0
go: downloading github.com/pires/go-proxyproto v0.8.1
go: downloading github.com/tailscale/wireguard-go v0.0.0-20250716170648-1d0488a3d7da
go: downloading github.com/jellydator/ttlcache/v3 v3.1.0
go: downloading github.com/tailscale/peercred v0.0.0-20250107143737-35a0c7bd7edc
go: downloading github.com/mdlayher/genetlink v1.3.2
go: downloading github.com/illarion/gonotify/v3 v3.0.2
go: downloading github.com/mdlayher/sdnotify v1.0.0
go: downloading github.com/insomniacslk/dhcp v0.0.0-20231206064809-8c70d406f6d2
go: downloading github.com/google/go-tpm v0.9.4
go: downloading github.com/tailscale/netlink v1.1.1-0.20240822203006-4d49adab4de7
go: downloading github.com/kortschak/wol v0.0.0-20200729010619-da482cc4850a
go: downloading github.com/anmitsu/go-shlex v0.0.0-20200514113438-38f4b401e2be
go: downloading github.com/kr/fs v0.1.0
go: downloading github.com/safchain/ethtool v0.3.0
go: downloading github.com/digitalocean/go-smbios v0.0.0-20180907143718-390a4f403a8e
go: downloading github.com/u-root/uio v0.0.0-20240224005618-d2acac8f3701
go: downloading github.com/coreos/go-iptables v0.7.1-0.20240112124308-65c67c9f46e6
go: downloading github.com/google/nftables v0.2.1-0.20240414091927-5e242ec57806
go: downloading github.com/pierrec/lz4/v4 v4.1.25
2026/03/06 04:28:40 
go: downloading github.com/tailscale/gokrazy-kernel v0.0.0-20240728225134-3d23beabda2e
2026/03/06 04:28:41 Including extra files for Go packages:
2026/03/06 04:28:41 
2026/03/06 04:28:41   tailscale.com/cmd/tailscale
2026/03/06 04:28:41     will include extra files in the root file system
2026/03/06 04:28:41       from usr-dir.tar
2026/03/06 04:28:41       last modified: 2024-06-03T04:59:03Z (15383h29m39s ago)
2026/03/06 04:28:41 
2026/03/06 04:28:42 
2026/03/06 04:28:42 Feature summary:
2026/03/06 04:28:42   use GPT: true
2026/03/06 04:28:42   use PARTUUID: true
2026/03/06 04:28:42   use GPT PARTUUID: true
2026/03/06 04:28:42 
2026/03/06 04:28:42 Creating boot file system
2026/03/06 04:28:42 
2026/03/06 04:28:42 Kernel directory: /home/runner/go/pkg/mod/github.com/tailscale/gokrazy-kernel@v0.0.0-20240728225134-3d23beabda2e
2026/03/06 04:28:42 MBR summary:
2026/03/06 04:28:42   LBAs: vmlinuz=8224 cmdline.txt=34644
2026/03/06 04:28:42   PARTUUID: 76baa2d6
2026/03/06 04:28:42 
2026/03/06 04:28:42 Creating root file system
If your applications need to store persistent data, create a file system using e.g.:
	/sbin/mkfs.ext4 -F -E offset=1157627904 /home/runner/work/tailscale/tailscale/gokrazy/natlabapp.img 98295

2026/03/06 04:28:43 To boot gokrazy, copy /home/runner/work/tailscale/tailscale/gokrazy/natlabapp.img to an SD card and plug it into a supported device (see https://gokrazy.org/platforms/)
2026/03/06 04:28:43 
2026/03/06 04:28:43 
2026/03/06 04:28:43 Build complete!
2026/03/06 04:28:43 
2026/03/06 04:28:43 To interact with the device, gokrazy provides a web interface reachable at:
2026/03/06 04:28:43 
2026/03/06 04:28:43 	***natlabapp/
2026/03/06 04:28:43 
2026/03/06 04:28:43 In addition, the following Linux consoles are set up:
2026/03/06 04:28:43 
2026/03/06 04:28:43 	1. foreground Linux console on the serial port (115200n8)
2026/03/06 04:28:43 	2. secondary Linux framebuffer console on HDMI
2026/03/06 04:28:43 
2026/03/06 04:28:43 Running /sbin/mkfs.ext4 ["-F" "-E" "offset=1157627904" "/home/runner/work/tailscale/tailscale/gokrazy/natlabapp.img" "98295"] ...
2026/03/06 04:28:43 Success.
2026/03/06 04:28:43 built. stopping.
qemu-img convert -O qcow2 natlabapp.img natlabapp.qcow2
    nat_test.go:89: found kernel: /home/runner/go/pkg/mod/github.com/tailscale/gokrazy-kernel@v0.0.0-20240728225134-3d23beabda2e/vmlinuz
    nat_test.go:397: node2 calling Status...
    nat_test.go:397: node1 calling Status...
2026/03/06 04:28:43 Got conn *net.UnixConn 0x2b9245820b8
2026/03/06 04:28:43 Got conn *net.UnixConn 0x2b9245820c0
[    1.529968] random: crng init done
[    1.533361] Linux agpgart interface v0.103
[    1.535345] ACPI: bus type drm_connector registered
[    1.550929] loop: module loaded
[    1.551942] virtio_blk virtio0: 1/0/0 default/read/poll queues
[    1.555112] virtio_blk virtio0: [vda] 2457616 512-byte logical blocks (1.26 GB/1.17 GiB)
[    1.570128]  vda: vda1 vda2 vda3 vda4
[    1.582304] wireguard: WireGuard 1.0.0 loaded. See www.wireguard.com for information.
[    1.582473] wireguard: Copyright (C) 2015-2019 Jason A. Donenfeld <Jason@zx2c4.com>. All Rights Reserved.
[    1.586514] tun: Universal TUN/TAP device driver, 1.6
[    1.592937] e100: Intel(R) PRO/100 Network Driver
[    1.593322] e100: Copyright(c) 1999-2006 Intel Corporation
[    1.593519] e1000: Intel(R) PRO/1000 Network Driver
[    1.593986] e1000: Copyright (c) 1999-2006 Intel Corporation.
[    1.594609] e1000e: Intel(R) PRO/1000 Network Driver
[    1.595472] e1000e: Copyright(c) 1999 - 2015 Intel Corporation.
[    1.596492] sky2: driver version 1.30
[    1.600093] usbcore: registered new interface driver usblp
[    1.600671] usbcore: registered new interface driver usb-storage
[    1.601618] i8042: PNP: No PS/2 controller found.
[    1.606185] rtc_cmos 00:00: registered as rtc0
[    1.607489] rtc_cmos 00:00: alarms up to one day, 242 bytes nvram
[    1.608229] fail to initialize ptp_kvm
[    1.608862] device-mapper: ioctl: 4.48.0-ioctl (2023-03-01) initialised: dm-devel@redhat.com
[    1.611109] hid: raw HID events driver (C) Jiri Kosina
[    1.613759] usbcore: registered new interface driver usbhid
[    1.614143] usbhid: USB HID core driver
[    1.626334] Initializing XFRM netlink socket
[    1.626870] NET: Registered PF_INET6 protocol family
[    1.633860] Segment Routing with IPv6
[    1.634351] In-situ OAM (IOAM) with IPv6
[    1.635750] sit: IPv6, IPv4 and MPLS over IPv4 tunneling driver
[    1.639644] NET: Registered PF_PACKET protocol family
[    1.641809] 9pnet: Installing 9P2000 support
[    1.642449] Key type dns_resolver registered
[    1.645631] IPI shorthand broadcast: enabled
[    1.682473] registered taskstats version 1
[    1.683551] Loading compiled-in X.509 certificates
[    1.696534] PM:   Magic number: 2:778:463
[    1.697299] printk: legacy console [netcon0] enabled
[    1.697584] netconsole: network logging started
[    1.698679] powernow_k8: Power state transitions not supported
[    1.700272] cfg80211: Loading compiled-in X.509 certificates for regulatory database
[    1.708906] kworker/u2:1 (50) used greatest stack depth: 14496 bytes left
[    1.719276] Loaded X.509 cert 'sforshee: 00b28ddf47aef9cea7'
[    1.720455] Loaded X.509 cert 'wens: 61c038651aabdcf94bd0ac7ff06c7248db18c600'
[    1.722918] Unstable clock detected, switching default tracing clock to "global"
[    1.722918] If you want to keep using the local clock, then add:
[    1.722918]   "trace_clock=local"
[    1.722918] on the kernel command line
[    1.725081] platform regulatory.0: Direct firmware load for regulatory.db failed with error -2
[    1.726648] cfg80211: failed to load regulatory.db
[    1.728899] clk: Disabling unused clocks
[    1.729516] ALSA device list:
[    1.729930]   No soundcards found.
[    1.736146] md: Waiting for all devices to be available before autodetect
[    1.736487] md: If you don't use raid, use raid=noautodetect
[    1.737126] md: Autodetecting RAID arrays.
[    1.737518] md: autorun ...
[    1.737925] md: ... autorun DONE.
[    1.753743] VFS: Mounted root (squashfs filesystem) readonly on device 253:2.
[    1.755591] devtmpfs: mounted
[    1.827160] Freeing unused kernel image (initmem) memory: 3104K
[    1.827578] Write protecting the kernel read-only data: 28672k
[    1.829254] Freeing unused kernel image (rodata/data gap) memory: 68K
[    1.992315] x86/mm: Checked W+X mappings: passed, no W+X pages found.
[    1.993820] Run /gokrazy/init as init process
gokrazy build timestamp 2026-03-06T04:27:15Z
2026/03/06 04:28:45 gokrazy.go:62: disabling hardware watchdog, as it could not be opened: open /dev/watchdog: no such file or directory
[    2.470861] EXT4-fs (vda4): mounted filesystem 5e43fc3c-4cfa-4224-8e0e-cef8d6e5c350 r/w with ordered data mode. Quota mode: none.
2026/03/06 04:28:45 gokrazy.go:181: cmdline: console=tty1 console=ttyS0,115200 root=PARTUUID=60c24cc1-f3f9-427a-8199-76baa2d60001/PARTNROFF=1 ro init=/gokrazy/init panic=10 oops=panic                                                                
2026/03/06 04:28:45 gokrazy.go:233: loading modules: open /lib/modules/6.8.9/kernel/drivers/hwmon/pwm-fan.ko: no such file or directory
hostname "natlabapp"
gokrazy device model Bochs
2026/03/06 04:28:45 eeprom.go:28: (not critical) could not read running EEPROM firmware version: open /sys/firmware/devicetree/base/chosen/bootloader/build-timestamp: no such file or directory
2026/03/06 04:28:45 gokrazy.go:450: getting EEPROM version: "/pieeprom.sig" not found
2026/03/06 04:28:45 powerbutton.go:22: polling device /dev/input/event0 for power button events
2026/03/06 04:28:45 update.go:311: update handler will be called with inactiveRootPartition = 3
2026/03/06 04:28:45 listeners.go:110: now listening on /run/gokrazy-http.sock
[    3.856718] serial-busybox (67) used greatest stack depth: 13704 bytes left
2026/03/06 04:28:49 [conn 0x2b9245820c0] Registering writer for MAC 52:cc:cc:cc:cc:01, node 192.168.1.101
2026/03/06 04:28:49 [net-52:ee:ee:ee:ee:01] dropping IPv6 packet on v4-only network
2026/03/06 04:28:49 vnet-AcceptTCP: 192.168.1.101:48128 -> 52.52.0.2:8008
2026/03/06 04:28:49 vnet-AcceptTCP: 192.168.1.101:48144 -> 52.52.0.2:8008
2026/03/06 04:28:49 vnet-AcceptTCP: 192.168.1.101:48160 -> 52.52.0.2:8008
2026/03/06 04:28:50 [net-52:ee:ee:ee:ee:01] dropping IPv6 packet on v4-only network
2026/03/06 04:28:50 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:28:50 [net-52:ee:ee:ee:ee:01] dropping IPv6 packet on v4-only network
    nat_test.go:402: node1 status: NeedsLogin
2026/03/06 04:28:50 vnet-AcceptTCP: 192.168.1.101:48174 -> 52.52.0.2:8008
2026/03/06 04:28:50 vnet-AcceptTCP: 192.168.1.101:60316 -> 52.52.0.3:80
2026/03/06 04:28:50 vnet-AcceptTCP: 192.168.1.101:60332 -> 52.52.0.3:80
2026/03/06 04:28:51 Created user userid:1 for node nodekey:90e254482cae8bae46e85329e6e3321f952e3e0842d76c26612f5714a937e050
2026/03/06 04:28:51 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:28:51 [net-52:ee:ee:ee:ee:01] dropping IPv6 packet on v4-only network
2026/03/06 04:28:51 [net-52:ee:ee:ee:ee:01] dropping IPv6 packet on v4-only network
2026/03/06 04:28:51 [net-52:ee:ee:ee:ee:01] dropping IPv6 packet on v4-only network
2026/03/06 04:28:51 vnet-AcceptTCP: 192.168.1.101:34554 -> 33.4.0.1:443
2026/03/06 04:28:51 http: TLS handshake error from 192.168.1.101:34554: EOF
2026/03/06 04:28:51 vnet-AcceptTCP: 192.168.1.101:34566 -> 33.4.0.1:443
    nat_test.go:414: node1 up!
    nat_test.go:425: node1 up with [100.64.0.1 fd7a:115c:a1e0:ab12:4843:cd96:6240:1]
2026/03/06 04:28:51 vnet-AcceptTCP: 192.168.1.101:44506 -> 33.4.0.2:443
2026/03/06 04:28:52 vnet-AcceptTCP: 192.168.1.101:36840 -> 52.52.0.4:443
2026/03/06 04:28:52 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:28:53 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:28:54 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:28:55 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:28:55 [net-52:ee:ee:ee:ee:01] dropping IPv6 packet on v4-only network
2026/03/06 04:28:56 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:28:57 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:28:58 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:28:59 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:00 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:01 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:02 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:03 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:03 [net-52:ee:ee:ee:ee:01] dropping IPv6 packet on v4-only network
2026/03/06 04:29:04 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:05 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:06 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:07 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:08 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:09 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:10 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:11 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:12 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:13 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:14 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:15 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:16 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:17 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:18 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:19 [net-52:ee:ee:ee:ee:01] dropping IPv6 packet on v4-only network
2026/03/06 04:29:19 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:20 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:21 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:22 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:23 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:24 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:25 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:26 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:27 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:28 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:29 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:30 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:31 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:32 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:33 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:34 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:35 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:36 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:37 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:38 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:39 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:40 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:41 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
2026/03/06 04:29:42 takeAgentConnOne: missed 2 times for 52:cc:cc:cc:cc:02
    nat_test.go:430: initial setup: node2 status: Get "http://local-tailscaled.sock/localapi/v0/status": context deadline exceeded
2026/03/06 04:29:43 ReadFull header: EOF
2026/03/06 04:29:43 ReadFull header: EOF
--- FAIL: TestEasyEasy (150.47s)
FAIL
FAIL	tailscale.com/tstest/integration/nat	150.480s
FAIL
```

cc @cmol

--- a/.github/workflows/natlab-integrationtest.yml
+++ b/.github/workflows/natlab-integrationtest.yml
@@ -22,12 +22,23 @@ jobs:
     steps:
       - name: Check out code
         uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6.0.2
+      - name: Enable KVM
+        run: |
+          echo 'KERNEL=="kvm", GROUP="kvm", MODE="0666", OPTIONS+="static_node=kvm"' | sudo tee /etc/udev/rules.d/99-kvm4all.rules
+          sudo udevadm control --reload-rules
+          sudo udevadm trigger --name-match=kvm
       - name: Install qemu
         run: |
           sudo rm -f /var/lib/man-db/auto-update
           sudo apt-get -y update
           sudo apt-get -y remove man-db
           sudo apt-get install -y qemu-system-x86 qemu-utils
+      - name: Build VM image
+        # The test will build this if missing, but we do it explicitly
+        # to avoid cutting into the go test -timeout budget, and to
+        # fail earlier with a clearer error if the image build breaks.
+        run: |
+          make -C gokrazy natlab
       - name: Run natlab integration tests
         run: |
           ./tool/go test -v -run=^TestEasyEasy$ -timeout=3m -count=1 ./tstest/integration/nat  --run-vm-tests

--- a/tstest/integration/nat/nat_test.go
+++ b/tstest/integration/nat/nat_test.go
@@ -7,6 +7,7 @@ import (
 	"bytes"
 	"cmp"
 	"context"
+	"encoding/json"
 	"errors"
 	"flag"
 	"fmt"
@@ -17,6 +18,7 @@ import (
 	"os"
 	"os/exec"
 	"path/filepath"
+	"runtime"
 	"strconv"
 	"strings"
 	"sync"
@@ -327,6 +329,15 @@ func (nt *natTest) setupTest(ctx context.Context, addNode ...addNodeFunc) (nodes
 		}
 	})
 
+	haveKVM := false
+	if runtime.GOOS == "linux" {
+		if f, err := os.OpenFile("/dev/kvm", os.O_RDWR, 0); err == nil {
+			f.Close()
+			haveKVM = true
+		}
+	}
+
+	qmpSocks := make([]string, len(nodes))
 	for i, node := range nodes {
 		disk := fmt.Sprintf("%s/node-%d.qcow2", nt.tempDir, i)
 		out, err := exec.Command("qemu-img", "create",
@@ -349,22 +360,28 @@ func (nt *natTest) setupTest(ctx context.Context, addNode ...addNodeFunc) (nodes
 		}
 		envStr := envBuf.String()
 
-		cmd := exec.Command("qemu-system-x86_64",
+		qmpSocks[i] = fmt.Sprintf("%s/qmp-node-%d.sock", nt.tempDir, i)
+		qemuArgs := []string{
 			"-M", "microvm,isa-serial=off",
 			"-m", "384M",
 			"-nodefaults", "-no-user-config", "-nographic",
 			"-kernel", nt.kernel,
-			"-append", "console=hvc0 root=PARTUUID=60c24cc1-f3f9-427a-8199-76baa2d60001/PARTNROFF=1 ro init=/gokrazy/init panic=10 oops=panic pci=off nousb tsc=unstable clocksource=hpet gokrazy.remote_syslog.target="+sysLogAddr+" tailscale-tta=1"+envStr,
-			"-drive", "id=blk0,file="+disk+",format=qcow2",
+			"-append", "console=hvc0 root=PARTUUID=60c24cc1-f3f9-427a-8199-76baa2d60001/PARTNROFF=1 ro init=/gokrazy/init panic=10 oops=panic pci=off nousb gokrazy.remote_syslog.target=" + sysLogAddr + " tailscale-tta=1" + envStr,
+			"-drive", "id=blk0,file=" + disk + ",format=qcow2",
 			"-device", "virtio-blk-device,drive=blk0",
-			"-netdev", "stream,id=net0,addr.type=unix,addr.path="+sockAddr,
+			"-netdev", "stream,id=net0,addr.type=unix,addr.path=" + sockAddr,
 			"-device", "virtio-serial-device",
 			"-device", "virtio-rng-device",
-			"-device", "virtio-net-device,netdev=net0,mac="+node.MAC().String(),
+			"-device", "virtio-net-device,netdev=net0,mac=" + node.MAC().String(),
 			"-chardev", "stdio,id=virtiocon0,mux=on",
 			"-device", "virtconsole,chardev=virtiocon0",
 			"-mon", "chardev=virtiocon0,mode=readline",
-		)
+			"-qmp", "unix:" + qmpSocks[i] + ",server=on,wait=off",
+		}
+		if haveKVM {
+			qemuArgs = append(qemuArgs, "-enable-kvm", "-cpu", "host")
+		}
+		cmd := exec.Command("qemu-system-x86_64", qemuArgs...)
 		cmd.Stdout = os.Stdout
 		cmd.Stderr = os.Stderr
 		if err := cmd.Start(); err != nil {
@@ -374,6 +391,15 @@ func (nt *natTest) setupTest(ctx context.Context, addNode ...addNodeFunc) (nodes
 			cmd.Process.Kill()
 			cmd.Wait()
 		})
+	}
+
+	for i, node := range nodes {
+		if err := nt.vnet.AwaitFirstPacket(ctx, node.MAC()); err != nil {
+			t.Logf("node %v: no boot progress (no packets received): %v", node, err)
+			t.Logf("node %v: QMP status: %s", node, qmpQueryStatus(qmpSocks[i]))
+			t.FailNow()
+		}
+		t.Logf("node %v: boot detected (first packet received)", node)
 	}
 
 	for _, n := range nodes {
@@ -430,8 +456,24 @@ func (nt *natTest) setupTest(ctx context.Context, addNode ...addNodeFunc) (nodes
 	return nodes, clients, nt.vnet.Close
 }
 
+type hasDeadline interface {
+	Deadline() (deadline time.Time, ok bool)
+}
+
+// testContext returns a context derived from the test's deadline (from -timeout),
+// leaving a small margin for cleanup. Falls back to 60s if no deadline is set.
+func testContext(tb testing.TB) (context.Context, context.CancelFunc) {
+	if t, ok := tb.(hasDeadline); ok {
+		if dl, ok := t.Deadline(); ok {
+			const margin = 5 * time.Second
+			return context.WithDeadline(context.Background(), dl.Add(-margin))
+		}
+	}
+	return context.WithTimeout(context.Background(), 60*time.Second)
+}
+
 func (nt *natTest) runHostConnectivityTest(addNode ...addNodeFunc) bool {
-	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	ctx, cancel := testContext(nt.tb)
 	defer cancel()
 	nodes, clients, cleanup := nt.setupTest(ctx, addNode...)
 	defer cleanup()
@@ -457,7 +499,7 @@ func (nt *natTest) runHostConnectivityTest(addNode ...addNodeFunc) bool {
 }
 
 func (nt *natTest) runTailscaleConnectivityTest(addNode ...addNodeFunc) pingRoute {
-	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	ctx, cancel := testContext(nt.tb)
 	defer cancel()
 
 	nodes, clients, cleanup := nt.setupTest(ctx, addNode...)
@@ -578,6 +620,55 @@ func ping(ctx context.Context, t testing.TB, c *vnet.NodeAgentClient, target net
 		return lastRes, nil
 	}
 	return nil, fmt.Errorf("no ping response (ctx: %v)", ctx.Err())
+}
+
+// qmpQueryStatus connects to a QEMU QMP socket and returns the VM status
+// (e.g. "running", "paused", "prelaunch") or an error string.
+func qmpQueryStatus(sockPath string) string {
+	conn, err := net.DialTimeout("unix", sockPath, 2*time.Second)
+	if err != nil {
+		return fmt.Sprintf("dial error: %v", err)
+	}
+	defer conn.Close()
+	conn.SetDeadline(time.Now().Add(5 * time.Second))
+	dec := json.NewDecoder(conn)
+
+	// Read QMP greeting.
+	var greeting json.RawMessage
+	if err := dec.Decode(&greeting); err != nil {
+		return fmt.Sprintf("greeting error: %v", err)
+	}
+
+	// Enter command mode.
+	if _, err := conn.Write([]byte(`{"execute":"qmp_capabilities"}` + "\n")); err != nil {
+		return fmt.Sprintf("write caps: %v", err)
+	}
+	var capsResp json.RawMessage
+	if err := dec.Decode(&capsResp); err != nil {
+		return fmt.Sprintf("caps response: %v", err)
+	}
+
+	// Query status.
+	if _, err := conn.Write([]byte(`{"execute":"query-status"}` + "\n")); err != nil {
+		return fmt.Sprintf("write query-status: %v", err)
+	}
+	var statusResp struct {
+		Return struct {
+			Running bool   `json:"running"`
+			Status  string `json:"status"`
+		} `json:"return"`
+		Error *struct {
+			Class string `json:"class"`
+			Desc  string `json:"desc"`
+		} `json:"error"`
+	}
+	if err := dec.Decode(&statusResp); err != nil {
+		return fmt.Sprintf("status response: %v", err)
+	}
+	if statusResp.Error != nil {
+		return fmt.Sprintf("qmp error: %s: %s", statusResp.Error.Class, statusResp.Error.Desc)
+	}
+	return fmt.Sprintf("status=%s running=%v", statusResp.Return.Status, statusResp.Return.Running)
 }
 
 func up(ctx context.Context, c *vnet.NodeAgentClient) error {

--- a/tstest/natlab/vnet/vnet.go
+++ b/tstest/natlab/vnet/vnet.go
@@ -747,6 +747,7 @@ type Server struct {
 	agentConnWaiter map[*node]chan<- struct{} // signaled after added to set
 	agentConns      set.Set[*agentConn]       //  not keyed by node; should be small/cheap enough to scan all
 	agentDialer     map[*node]netx.DialFunc
+	gotFirstPacket  map[MAC]chan struct{} // closed on first packet from each MAC
 
 	cloudInitData map[int]*CloudInitData // node num → cloud-init config
 	fileContents  map[string][]byte      // filename → file bytes
@@ -824,6 +825,10 @@ func New(c *Config) (*Server, error) {
 	}
 	if err := s.initFromConfig(c); err != nil {
 		return nil, err
+	}
+	s.gotFirstPacket = make(map[MAC]chan struct{})
+	for mac := range s.nodeByMAC {
+		s.gotFirstPacket[mac] = make(chan struct{})
 	}
 	for n := range s.networks {
 		if err := n.initStack(); err != nil {
@@ -930,6 +935,22 @@ func (s *Server) Close() {
 		s.pcapWriter.Close()
 	}
 	s.wg.Wait()
+}
+
+// AwaitFirstPacket waits until the first ethernet frame is received from the
+// given MAC address, indicating the VM has booted far enough to send network
+// traffic. It returns an error if the context expires first.
+func (s *Server) AwaitFirstPacket(ctx context.Context, mac MAC) error {
+	ch, ok := s.gotFirstPacket[mac]
+	if !ok {
+		return fmt.Errorf("unknown MAC %v", mac)
+	}
+	select {
+	case <-ch:
+		return nil
+	case <-ctx.Done():
+		return fmt.Errorf("no network packets received from %v: %w", mac, ctx.Err())
+	}
 }
 
 // MACs returns the MAC addresses of the configured nodes.
@@ -1088,6 +1109,13 @@ func (s *Server) ServeUnixConn(uc *net.UnixConn, proto Protocol) {
 		}
 		if !didReg[srcMAC] {
 			didReg[srcMAC] = true
+			if ch, ok := s.gotFirstPacket[srcMAC]; ok {
+				select {
+				case <-ch: // already closed
+				default:
+					close(ch)
+				}
+			}
 			srcNet := srcNode.netForMAC(srcMAC)
 			if srcNet == nil {
 				s.logf("[conn %p] node %v has no network for MAC %v", c.uc, srcNode, srcMAC)

